chia-protocol/src/program.rs

```diff
@@ -59,6 +50,257 @@ impl Program {
+        args: &PyAny,
+    ) -> PyResult<(u64, &'a PyAny)> {
+        use clvmr::MEMPOOL_MODE;
+        self._run(py, max_cost, MEMPOOL_MODE, args)
+    }
+
+    fn run_with_cost<'a>(
+        &self,
+        py: Python<'a>,
+        max_cost: u64,
+        args: &PyAny,
+    ) -> PyResult<(u64, &'a PyAny)> {
+        self._run(py, max_cost, 0, args)
+    }
+
+    fn _run<'a>(
+        &self,
+        py: Python<'a>,
+        max_cost: u64,
+        flags: u32,
+        args: &PyAny,
+    ) -> PyResult<(u64, &'a PyAny)> {
+        use clvmr::reduction::Response;
+        use clvmr::run_program;
+        use clvmr::ChiaDialect;
+        use std::rc::Rc;
+
+        let mut a = Allocator::new_limited(500000000, 62500000, 62500000);
+        let clvm_args = clvm_convert(&mut a, args)?;
+
+        let r: Response = (|| -> PyResult<Response> {
+            let program = node_from_bytes_backrefs(&mut a, self.0.as_ref())?;
+            let dialect = ChiaDialect::new(flags);
+
+            Ok(py.allow_threads(|| run_program(&mut a, &dialect, program, clvm_args, max_cost)))
+        })()?;
+        match r {
+            Ok(reduction) => {
+                let val = LazyNode::new(Rc::new(a), reduction.1);
+                Ok((reduction.0, to_program(py, val)?))
+            }
+            Err(eval_err) => Err(PyValueError::new_err(eval_err.to_string())),
+        }
+    }
+
+    fn to_program<'a>(&self, py: Python<'a>) -> PyResult<&'a PyAny> {
+        use std::rc::Rc;
+        let mut a = Allocator::new_limited(500000000, 62500000, 62500000);
+        let prg = node_from_bytes_backrefs(&mut a, self.0.as_ref())?;
+        let prg = LazyNode::new(Rc::new(a), prg);
+        to_program(py, prg)
+    }
+
+    fn uncurry<'a>(&self, py: Python<'a>) -> PyResult<(&'a PyAny, &'a PyAny)> {
+        use clvm_utils::CurriedProgram;
+        use std::rc::Rc;
+
+        let mut a = Allocator::new_limited(500000000, 62500000, 62500000);
+        let prg = node_from_bytes_backrefs(&mut a, self.0.as_ref())?;
+        let Ok(uncurried) = CurriedProgram::<NodePtr>::from_clvm(&a, prg) else {
+            let a = Rc::new(a);
+            let prg = LazyNode::new(a.clone(), prg);
+            let ret = a.null();
+            let ret = LazyNode::new(a, ret);
+            return Ok((to_program(py, prg)?, to_program(py, ret)?));
+        };
+
+        let mut curried_args = Vec::<NodePtr>::new();
+        let mut args = uncurried.args;
+        loop {
+            if let SExp::Atom = a.sexp(args) {
+                break;
+            }
+            // the args of curried puzzles are in the form of:
+            // (c . ((q . <arg1>) . (<rest> . ())))
+            let (_, ((_, arg), (rest, _))) =
+                <(
+                    clvm_traits::MatchByte<4>,
+                    (clvm_traits::match_quote!(NodePtr), (NodePtr, ())),
+                ) as clvm_traits::FromClvm>::from_clvm(&a, args)?;
+            curried_args.push(arg);
+            args = rest;
+        }
```

Comment:
@Rigidity I couldn't find a better way of uncurrying an arbitrary number of arbitrary arguments. There's no trick to get `CurriedProgram` to do this, right?

Reply:
Not currently. I want to improve a bunch of the clvm-traits stuff in the future and make the code cleaner, but I don't have time right now.